onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder.cc

```diff
@@ -2694,6 +2694,108 @@
 
 #pragma endregion
 
+#pragma region op_pad
+
+class PadOpBuilder : public BaseOpBuilder {
+ public:
+  void AddInitializersToSkip(ModelBuilder& model_builder, const NodeUnit& node_unit) const override;
+
+ private:
+  Status AddToModelBuilderImpl(ModelBuilder& model_builder, const NodeUnit& node_unit) const override;
+};
+
+void PadOpBuilder::AddInitializersToSkip(ModelBuilder& model_builder, const NodeUnit& node_unit) const {
+  const auto& inputs = node_unit.Inputs();
+  model_builder.AddInitializerToSkip(inputs[1].node_arg.Name());  // pads
+  if (inputs.size() > 2) {
+    model_builder.AddInitializerToSkip(inputs[2].node_arg.Name());  // constant_value
+  }
+}
+
+Status PadOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_builder, const NodeUnit& node_unit) const {
+  auto& shaper = model_builder.GetShaper();
+  const auto& operand_indices = model_builder.GetOperandIndices();
+  const auto& operand_types = model_builder.GetOperandTypes();
+  const auto& inputs = node_unit.Inputs();
+  const auto& outputs = node_unit.Outputs();
+
+  std::vector<uint32_t> input_indices{};
+
+  // `data` input
+  const auto& data = inputs[0].node_arg.Name();
+  input_indices.push_back(operand_indices.at(data));
+
+  // `pads` input
+  // convert from [begin_1, begin_2, ..., end_1, end_2, ...] to [begin_1, end_1, begin_2, end_2, ...]
+  // convert from int64_t to int32_t
+  const auto& data_shape = shaper[data];
+  const uint32_t data_rank = SafeInt<uint32_t>(data_shape.size());
+
+  const auto& pads = inputs[1].node_arg.Name();
+  const auto* pads_initializer = model_builder.GetConstantInitializer(pads);
+  ORT_RETURN_IF_NOT(pads_initializer, "pads must be a constant");
+  ORT_ENFORCE(pads_initializer->data_type() == ONNX_NAMESPACE::TensorProto_DataType_INT64);
```

Comment:
Should this be done in PadOpSupportChecker?

Reply:
replaced with a comment. my thinking was that it was preferable to put the assumption in code. would assert() be ok here?

Reply:
The issue with `assert` is that also creates ambiguity - i.e. not clear to a developer why it's an enforce in places and assert in others. Possibly fine if the assert is wrapped in a macro with naming along the lines of 'op support checker has already validated this'.

We probably need to do a big cleanup of the existing code to really draw a clean line between what the op support checker does and what the build does. otherwise you end up with unnecessary checks in both places as it's not clear who owns doing the check.

Reply:
ok, will just have a comment then